### ISSUE TITLE
Shapes are misplaced when using numbers as categoryarray values

### ISSUE THREAD

**Issue by perdodi** [plotly/plotly.js#5767]
If X axis `type` is `category`, and `categoryarray` contains numbers (string type), then shapes are horizontally misplaced, as `x0` and `x1` values in the `shape` are not considered as _elements_, but as _indices_ of the `categoryarray`.

Eg.

This one works perfectly:

```
<div id="tester"></div>
<script>
    var layout = {
      "xaxis": {
        "type": "category",
        "categoryarray": [
          "red", "green", "blue", "purple", "cyan", "yellow", "magenta"
        ],
      },
      "shapes": [
        {
          "type": "rect",
          "xref": "x",
          "yref": "y",
          "x0": "purple",
          "x1": "yellow",
          "y0": 10,
          "y1": 20,
          "fillcolor": "rgba(45,180,135,0.3)",
        }
      ]
    };

    var data = [
      {
        "x": [ "red", "blue", "cyan", "magenta" ],
        "y": [ 10, 20, 30, 40],
      },
    ];
    
    TESTER = document.getElementById('tester');
    Plotly.newPlot( TESTER, data, layout );
</script>
```

![image](https://user-images.githubusercontent.com/1175690/123436868-60911700-d5cf-11eb-8954-6c64298f8602.png)

But if I replace the color names to numbers:

```
  // in "categoryarray":
      "10", "20", "30", "40", "50", "60", "70"
...
  // in "shapes" this won't work:
      "x0": "40",
      "x1": "60",
...
 // in data:
    "x": [ "10", "30", "50", "70" ],

```

![image](https://user-images.githubusercontent.com/1175690/123436896-671f8e80-d5cf-11eb-8772-e82f7646e00e.png)

It will work only if I replace the category values to their indices in the `shapes`:
```
      "x0": 3,
      "x1": 5,
```

![image](https://user-images.githubusercontent.com/1175690/123436923-6d156f80-d5cf-11eb-9225-c0c6e20da3d4.png)

Note: if this is not intended to be a bug, but an undocumented feature, please document it, so I can rely on it later. 😃

**Comment by nicholas-esterer** [comment]
This seems to be due to the behaviour of the package `fast-isnumeric`, which is used in https://github.com/plotly/plotly.js/blob/2f17e117648e56ff3eebb5f55e57db522f4d9200/src/plots/cartesian/set_convert.js#L182
Indeed, in a `nodejs` console:
```javascript
> var isNumeric = require('fast-isnumeric');
> isNumeric(40)
true
> isNumeric("40")
true
```
This makes it so that `getCategoryIndex` is not called even when the value is a string that looks like a number (e.g., "40").

**Comment by nicholas-esterer** [comment]
also related: https://github.com/plotly/plotly.js/issues/5741

**Comment by alexcjohnson** [comment]
See also `autotypenumbers: 'strict'` https://github.com/plotly/plotly.js/pull/5240

**Comment by alexcjohnson** [comment]
(which is now the default for graphs made with plotly.py)

**Comment by nicholas-esterer** [comment]
Ok I see. Setting `autotypenumbers: "strict"` in `layout` or `layout.xaxis` (relevant to this use case) doesn't fix the bug, but to be fair the docs just mention "trace data" and not the coordinates of shapes (or images or annotations for that matter). But the expected behaviour is most likely to follow what the axis' `autotypesnumbers` setting says to do for shapes as well.

**Comment by alexcjohnson** [comment]
Yep, that sounds right - if an axis is in strict mode, that should apply to shapes, annotations, and images as well as traces. So that part of this issue can be considered a bug.
